src/Workspaces/Remote/Core/Services/AssetService.cs

```diff
@@ -0,0 +1,165 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Execution;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    /// <summary>
+    /// This service provide a way to get roslyn objects from checksum
+    /// 
+    /// TODO: change this service to workspace service
+    /// </summary>
+    internal class AssetService
+    {
+        private const int CleanupInterval = 3; // 3 minutes
+        private const int PurgeAfter = 30; // 30 minutes
+
+        // PREVIEW: unfortunately, I need dummy workspace since workspace services can be workspace specific
+        private static readonly Serializer s_serializer = new Serializer(new AdhocWorkspace(RoslynServices.HostServices, workspaceKind: "dummy").Services);
+
+        private readonly ConcurrentDictionary<int, AssetSource> _assetSources =
+            new ConcurrentDictionary<int, AssetSource>(concurrencyLevel: 4, capacity: 10);
+
+        private readonly ConcurrentDictionary<Checksum, Entry> _assets =
+            new ConcurrentDictionary<Checksum, Entry>(concurrencyLevel: 4, capacity: 10);
+
+        public AssetService()
+        {
+            Task.Factory.StartNew(CleanAssets, CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default).Unwrap();
+        }
+
+        private async Task CleanAssets()
+        {
+            var purgeAfterTimeSpan = TimeSpan.FromMinutes(PurgeAfter);
+            var cleanupIntervalTimeSpan = TimeSpan.FromMinutes(CleanupInterval);
+
+            while (true)
+            {
+                var current = DateTime.UtcNow;
+
+                foreach (var kvp in _assets.ToArray())
+                {
+                    if (current - kvp.Value.LastAccessed <= purgeAfterTimeSpan)
+                    {
+                        continue;
+                    }
+
+                    // If it fails, we'll just leave it in the asset pool.
+                    Entry entry;
+                    _assets.TryRemove(kvp.Key, out entry);
+                }
+
+                await Task.Delay(cleanupIntervalTimeSpan).ConfigureAwait(false);
+            }
+        }
+
+        public void Set(Checksum checksum, object value)
+        {
+            _assets.TryAdd(checksum, new Entry(value));
+        }
+
+        public async Task<T> GetAssetAsync<T>(Checksum checksum, CancellationToken cancellationToken)
+        {
+            Entry entry;
+            if (!_assets.TryGetValue(checksum, out entry))
+            {
+                // TODO: what happen if service doesn't come back. timeout?
+                await RequestAssetAsync(checksum, cancellationToken).ConfigureAwait(false);
+
+                if (!_assets.TryGetValue(checksum, out entry))
```

Comment:
I simply mean that there's a race in this code.  You have two overlapping operations.  One checks a cache, requests an item (which will put the item into the cache) and checks the cache again.  The other is going through the cache and deleting things.

I really don't like that.  I'd far prefer that hte request of the asset returns the data.  You can then know that you have it without having to have scary Contract.Fail calls.

In other words: design so that it can't fail, instead of having to hope that it won't fail.